### PR TITLE
juju.client.gocookies: new module

### PR DESCRIPTION
**Pull request by rogpeppe**
This provides support for reading and writing
the cookies file that the Go bakery code uses.

Not yet implemented: file locking and cookie merging.

### CODE REVIEW

File: juju/client/gocookies.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    '''A CookieJar implementation that reads and writes cookies
+    to the cookiejar format as understood by the Go package
+    github.com/juju/persistent-cookiejar.'''
+    def _really_load(self, f, filename, ignore_discard, ignore_expires):
+        '''Implement the _really_load method called by FileCookieJar
+        to implement the actual cookie loading'''
+        now = time.time()
+        '''This method is called by FileCookieJar to do the cookie loading'''
+        for c in map(_new_py_cookie, json.loads(f.read())):
+            print('loading cookie {}'.format(c))
```
Comment (fabricematrat) [inline juju/client/gocookies.py]: leftover ?

File: juju/client/gocookies.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    c = go_cookie
+    expires = None
+    if c.get('Expires') is not None:
+        t = pyrfc3339.parse(c['Expires'])
+        expires = t.strftime("%s")
+    return cookiejar.Cookie(
+        version=0,
+        name=c['Name'],
+        value=c['Value'],
+        port=None,
```
Comment (fabricematrat) [inline juju/client/gocookies.py]: Shall we do something for the port ?

File: juju/client/gocookies.py (hunk 1 above)
Comment (rogpeppe) [reply juju/client/gocookies.py]: Yup, removed.

File: juju/client/gocookies.py (hunk 2 above)
Comment (rogpeppe) [reply juju/client/gocookies.py]: I don't understand why ports are there at all - the cookie RFC doesn't provide support for port-specific cookies.

File: tests/unit/test_gocookies.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    ('http://foo.com/path/x', [
+        ('foo4', 'foo4-value'),
+        ('foo1', 'foo1-value'),
+        ('foo3', 'foo3-value'),
+    ]),
+]
+
+class TestGoCookieJar(unittest.TestCase):
+    def setUp(self):
+         self.dir = tempfile.mkdtemp()
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: You could just `self.addCleanup(shutil.rmtree, self.dir)` here, and avoid defining the tearDown method at all. This also has the advantage that cleanup is done close to the thing.

File: tests/unit/test_gocookies.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        shutil.rmtree(self.dir)
+
+    def test_readcookies(self):
+        jar = self.load_jar(cookie_content)
+        self.assert_jar_queries(jar, cookie_content_queries)
+
+    def test_roundtrip(self):
+        jar = self.load_jar(cookie_content)
+        filename2 = os.path.join(self.dir, 'cookies2')
+        jar.save(filename = filename2)
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: For kwargs in Python we don;t generally use spaces around =. So this is `jar.save(filename=filename2)`. I wonder why lint is not complaining.

File: tests/unit/test_gocookies.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        jar = GoCookieJar()
+        jar.load(filename = filename2)
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: Ditto.

File: tests/unit/test_gocookies.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                "Persistent": true,
+                "Secure": false,
+                "Updated": "2017-11-17T08:53:55.088822562Z",
+                "Value": "bar-value"
+            }
+        ]'''
+        jar = self.load_jar(content)
+        gotExpires = tuple(jar)[0].expires
+        wantExpires = pyrfc3339.parse('2345-11-15T18:16:08Z').timestamp()
+        self.assertEqual(gotExpires, wantExpires)
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: Please always assert expected, obtained (so, expected first, unlike Go).

File: tests/unit/test_gocookies.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        return jar
+
+    def assert_jar_queries(self, jar, queries):
+        '''Assert that all the given queries (see cookie_content_queries)
+        are satisfied OK when run on the given cookie jar
+        :param suite TestCase: the test suite
+        :param jar CookieJar: the cookie jar to query
+        :param queries: the queries to run.
+        '''
+        for q in queries:
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: Single letter name.

File: tests/unit/test_gocookies.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        filename = os.path.join(self.dir, 'cookies')
+        with open(filename, 'x') as f:
+            f.write(content)
+        jar = GoCookieJar()
+        jar.load(filename=filename)
+        return jar
+
+    def assert_jar_queries(self, jar, queries):
+        '''Assert that all the given queries (see cookie_content_queries)
+        are satisfied OK when run on the given cookie jar
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: satisfied OK?

Comment (frankban) [reply tests/unit/test_gocookies.py]: Also final dot.

File: tests/unit/test_gocookies.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+                "Name": "bar",
+                "Path": "/",
+                "Persistent": true,
+                "Secure": false,
+                "Updated": "2017-11-17T08:53:55.088822562Z",
+                "Value": "bar-value"
+            }
+        ]'''
+        jar = self.load_jar(content)
+        gotExpires = tuple(jar)[0].expires
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: Do not use mixedCase in python for variable_names.

File: tests/unit/test_gocookies.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+    def assert_jar_queries(self, jar, queries):
+        '''Assert that all the given queries (see cookie_content_queries)
+        are satisfied OK when run on the given cookie jar
+        :param suite TestCase: the test suite
+        :param jar CookieJar: the cookie jar to query
+        :param queries: the queries to run.
+        '''
+        for q in queries:
+            url, wantCookies = q
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: Mixed case.

File: tests/unit/test_gocookies.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+            req = urllib.request.Request(url)
+            jar.add_cookie_header(req)
+            # We can't use SimpleCookie to find out what cookies
+            # have been presented, because SimpleCookie
+            # only allows one cookie with a given name,
+            # so we naively parse the cookies ourselves, which
+            # is OK because we know we don't have to deal
+            # with any complex cases.
+    
+            cookieHeader = req.get_header('Cookie')
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: cookie_header

File: tests/unit/test_gocookies.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+            gotCookies = []
+            if cookieHeader is not None:
+                gotCookies = [
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: got_cookies

File: tests/unit/test_gocookies.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+                    tuple(part.split('='))
+                    for part in cookieHeader.split('; ')
+                ]
+                gotCookies.sort()
+            wantCookies = list(wantCookies)
+            wantCookies.sort()
+            self.assertEqual(gotCookies, wantCookies, msg='query {}; got {}; want {}'.format(url, gotCookies, wantCookies))
```
Comment (frankban) [inline tests/unit/test_gocookies.py]: self.assertEqual(want_cookies, got_cookies, ...

File: tests/unit/test_gocookies.py (hunk 6 above)
Comment (rogpeppe) [reply tests/unit/test_gocookies.py]: Not as far as I can see from all the examples I've found online, e.g. https://docs.python.org/3/library/unittest.html

File: juju/client/gocookies.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
+import datetime
+import json
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done

File: juju/client/gocookies.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+import http.cookiejar as cookiejar
+import time
+import pyrfc3339
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done. I didn't know about that convention.

File: juju/client/gocookies.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+class GoCookieJar(cookiejar.FileCookieJar):
+    '''A CookieJar implementation that reads and writes cookies
+    to the cookiejar format as understood by the Go package
+    github.com/juju/persistent-cookiejar.'''
+    def _really_load(self, f, filename, ignore_discard, ignore_expires):
+        '''Implement the _really_load method called by FileCookieJar
+        to implement the actual cookie loading'''
+        now = time.time()
+        '''This method is called by FileCookieJar to do the cookie loading'''
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: removed

File: juju/client/gocookies.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+                if not ignore_expires and cookie.is_expired(now):
+                    continue
+                go_cookies.append(_new_go_cookie(cookie))
+            f.write(json.dumps(go_cookies))
+
+
+def _new_py_cookie(go_cookie):
+    '''Convert a Go-style JSON-unmarshaled cookie
+    into a Python cookie'''
+    c = go_cookie
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: removed

File: juju/client/gocookies.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+                if not ignore_discard and cookie.discard:
+                    continue
+                if not ignore_expires and cookie.is_expired(now):
+                    continue
+                go_cookies.append(_new_go_cookie(cookie))
+            f.write(json.dumps(go_cookies))
+
+
+def _new_py_cookie(go_cookie):
+    '''Convert a Go-style JSON-unmarshaled cookie
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done

File: juju/client/gocookies.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        now = time.time()
+        '''This method is called by FileCookieJar to do the cookie loading'''
+        for c in map(_new_py_cookie, json.loads(f.read())):
+            print('loading cookie {}'.format(c))
+            if not ignore_expires and c.is_expired(now):
+                continue
+            self.set_cookie(c)
+
+    def save(self, filename=None, ignore_discard=False, ignore_expires=False):
+        '''Implement the FileCookieJar abstract method'''
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done

File: juju/client/gocookies.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+class GoCookieJar(cookiejar.FileCookieJar):
+    '''A CookieJar implementation that reads and writes cookies
+    to the cookiejar format as understood by the Go package
+    github.com/juju/persistent-cookiejar.'''
+    def _really_load(self, f, filename, ignore_discard, ignore_expires):
+        '''Implement the _really_load method called by FileCookieJar
+        to implement the actual cookie loading'''
+        now = time.time()
+        '''This method is called by FileCookieJar to do the cookie loading'''
+        for c in map(_new_py_cookie, json.loads(f.read())):
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done

File: juju/client/gocookies.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        '''Implement the FileCookieJar abstract method'''
+        if filename is None:
+            if self.filename is not None:
+                filename = self.filename
+            else:
+                raise ValueError(cookiejar.MISSING_FILENAME_TEXT)
+
+        # TODO: obtain file lock, read contents of file, and merge with
+        # current content.
+        with open(filename, "w") as f:
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: When cookie merging is implemented, there will be, I think, but moved out of the scope for the time being.

File: juju/client/gocookies.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+        discard=False,
+        comment=None,
+        comment_url=None,
+        rest=None,
+        rfc2109=False,
+    )
+
+
+def _new_go_cookie(c):
+    '''Convert a python cookie to the JSON-marshalable
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done

File: juju/client/gocookies.py [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+    Go-style cookie form.'''
+    # not done yet:
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply juju/client/gocookies.py]: done

File: tests/unit/test_gocookies.py [hunk 24]
```diff
@@ -0,0 +1,244 @@
+"""
+Tests for the gocookies code.
+"""
+import os
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply tests/unit/test_gocookies.py]: done

File: tests/unit/test_gocookies.py (hunk 9 above)
Comment (rogpeppe) [reply tests/unit/test_gocookies.py]: done

File: tests/unit/test_gocookies.py (hunk 10 above)
Comment (rogpeppe) [reply tests/unit/test_gocookies.py]: actually q was unneeded

File: tests/unit/test_gocookies.py (hunk 13 above)
Comment (rogpeppe) [reply tests/unit/test_gocookies.py]: nope; see above.

**Pull request merged**